src/blocks/kdeconnect.rs

```diff
@@ -256,6 +295,16 @@ impl Device {
             .map(|n| n.len())
             .ok()
     }
+
+    async fn network(&self) -> (Option<String>, Option<i32>) {
+        (
```

Comment:
I'd use `tokio::join!` here and have the function signature:

```rust
async fn network(&self) -> (Result<String>, Result<i32>)
```

Reply:
I don't get this. I used the same format as that in the battery one few lines above this. Can I leave it for someone to change it when they change both. Or do you mean something that's specifically different in the cellular network that would require me to do that? I've never used tokio, so I don't know what I'll be doing.

Reply:
I would actually change the battery block as well... We can pull of the async code up higher in the code and perform it all at the same time. This is what I'd change (in regard to improving async performance):


```diff
diff --git a/src/blocks/kdeconnect.rs b/src/blocks/kdeconnect.rs                                                                                                                                                                                                
index ffaffec7..a289279f 100644                                                                                                                                                                                                                                 
--- a/src/blocks/kdeconnect.rs                                                                                                                                                                                                                                  
+++ b/src/blocks/kdeconnect.rs                                                                                                                                                                                                                                  
@@ -48,6 +48,7 @@                                                                                                                                                                                                                                               
 //! - `phone`                                                                                                                                                                                                                                                  
 //! - `phone_disconnected`                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                
+use futures::TryFutureExt;                                                                                                                                                                                                                                     
 use tokio::sync::mpsc;                                                                                                                                                                                                                                         
 use zbus::dbus_proxy;                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                
@@ -112,8 +113,13 @@ pub async fn run(config: Config, mut api: CommonApi) -> Result<()> {                                                                                                                                                                       
                 values.insert("icon".into(), Value::icon(api.get_icon("phone")?));                                                                                                                                                                             
                 values.insert("connected".into(), Value::flag());                                                                                                                                                                                              
                                                                                                                                                                                                                                                                
-                let (level, charging) = device.battery().await;                                                                                                                                                                                                
-                if let Some(level) = level {                                                                                                                                                                                                                   
+                let (                                                                                                                                                                                                                                          
+                    (level, charging),                                                                                                                                                                                                                         
+                    (cellular_network_type, cellular_network_strength),                      
+                    notif_count,                                                                                              
+                ) = tokio::join!(device.battery(), device.network(), device.notifications());                                  
+                                                                                                                               
+                if let Ok(level) = level {      
                     values.insert("bat_charge".into(), Value::percents(level));                                               
                     values.insert(                                                                                             
                         "bat_icon".into(), 
@@ -137,23 +143,21 @@ pub async fn run(config: Config, mut api: CommonApi) -> Result<()> {                                      
                     }         
                 }                                    
                                               
-                let network = device.network().await;                                                                          
-                if let Some(net) = network.0 {                                                                                
-                    let network_strength = network.1.unwrap_or(0);                                                            
+                if let Ok(cellular_network_type) = cellular_network_type {                                                    
                     values.insert(                                                                                             
                         "network_icon".into(),                                                                                
                         Value::icon(api.get_icon("net_wireless")?),                                                           
                     );                        
                                                                                                                               
-                    if network_strength <= 0 {                                                                                
+                    if cellular_network_strength <= 0 {                                                                        
                         widget.state = State::Critical;                                                                       
                         values.insert("network".into(), Value::text("×".into()));                                              
                     } else {                                                                                                  
-                        values.insert("network".into(), Value::text(net));                                                    
+                        values.insert("network".into(), Value::text(cellular_network_type));                                   
                     }                     
                 }                                                                                                             
                                                            
-                let notif_count = device.notifications().await.unwrap_or(0);                                                   
+                let notif_count = notif_count.unwrap_or(0);                                                                    
                 if notif_count > 0 {                 
                     values.insert("notif_count".into(), Value::number(notif_count));                                          
                     values.insert(                                                                                            
@@ -277,14 +281,15 @@ impl Device {                                                                                            
         self.device_proxy.name().await.ok()
     }                                              
                                                                                                                               
-    async fn battery(&self) -> (Option<u8>, bool) {
-        (            
+    async fn battery(&self) -> (Result<u8>, bool) {
+        tokio::join!(                                  
             self.battery_proxy                         
                 .charge()                                                                                                     
-                .await                          
-                .ok()                                                                                                         
-                .map(|p| p.clamp(0, 100) as u8),                                                                              
-            self.battery_proxy.is_charging().await.unwrap_or(false),                                                           
+                .map_ok(|p| p.clamp(0, 100) as u8)
+                .map_err(|_| Error::new("Could not get charge")),                                                              
+            self.battery_proxy                                                                                                
+                .is_charging()                             
+                .map_ok_or_else(|_| false, |x| x),
         )                                                                                                                     
     }                             
                                   
@@ -296,13 +301,14 @@ impl Device {         
             .ok()
     }                                                                                                                          
                                                    
-    async fn network(&self) -> (Option<String>, Option<i32>) {                                                                 
-        (                                             
-            self.connectivity_proxy.cellular_network_type().await.ok(),                                                        
+    async fn network(&self) -> (Result<String>, i32) {
+        tokio::join!(                                          
+            self.connectivity_proxy                                                                                            
+                .cellular_network_type()                       
+                .map_err(|_| Error::new("Could not get cellular_network_type")),                                               
             self.connectivity_proxy                            
                 .cellular_network_strength()                   
-                .await                                         
-                .ok(),                                         
+                .map_ok_or_else(|_| 0, |x| x),                 
         )                                                      
     }                                                                                                                          
 }             
```